### PR TITLE
[prometheus-blackbox-exporter] remove helm2 support, ingress update, fix docs & use label template

### PR DESCRIPTION
**Pull request by monotek**
Signed-off-by: André Bauer <monotek23@gmail.com>

#### What this PR does / why we need it:
* remove helm 2 support
* ingress update (adds ingressClassName & pathType)
* fix docs (extraArgs)
* use label template

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #
https://github.com/prometheus-community/helm-charts/issues/1111

#### Special notes for your reviewer:
* updates github actions for chart testing & kind 
* updates helm version to 3.6.2

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Review by torstenwalter — APPROVED** [review]

File: charts/prometheus-blackbox-exporter/templates/ingress.yaml [hunk 1]
```diff
@@ -1,35 +1,61 @@
...
-    app.kubernetes.io/managed-by: {{ .Release.Service }}
-    helm.sh/chart: {{ include "prometheus-blackbox-exporter.chart" . }}
+    {{- include "prometheus-blackbox-exporter.labels" . | nindent 4 }}
+  {{- with .Values.ingress.annotations }}
   annotations:
-{{ toYaml .Values.ingress.annotations | indent 4 }}
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
 spec:
+  {{- if and .Values.ingress.className (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) }}
```
Comment (desaintmartin) [inline charts/prometheus-blackbox-exporter/templates/ingress.yaml:26]: Should we fail if `.Values.ingress.className` but kube < 1.18?

File: charts/prometheus-blackbox-exporter/templates/ingress.yaml [hunk 2]
```diff
@@ -1,35 +1,61 @@
...
+  ingressClassName: {{ .Values.ingress.className }}
+  {{- end }}
+  {{- if .Values.ingress.tls }}
```
Comment (desaintmartin) [inline charts/prometheus-blackbox-exporter/templates/ingress.yaml:29]: `with`?

File: charts/prometheus-blackbox-exporter/templates/service.yaml [hunk 3]
```diff
@@ -1,30 +1,19 @@
-kind: Service
 apiVersion: v1
+kind: Service
 metadata:
-  name: {{ template "prometheus-blackbox-exporter.fullname" . }}
-  {{- if .Values.service.annotations }}
-  annotations:
```
Comment (desaintmartin) [inline charts/prometheus-blackbox-exporter/templates/service.yaml:6]: Was this removal intentional?

File: charts/prometheus-blackbox-exporter/templates/service.yaml [hunk 4]
```diff
@@ -1,30 +1,19 @@
...
-    - name: http
-      port: {{ .Values.service.port }}
+    - port: {{ .Values.service.port }}
       targetPort: http
       protocol: TCP
+      name: http
 {{- if .Values.service.externalIPs }}
   externalIPs:
 {{ toYaml .Values.service.externalIPs | indent 4 }}
-{{- end }}
```
Comment (desaintmartin) [inline charts/prometheus-blackbox-exporter/templates/service.yaml:27]: Shouldn't we keep it? Well, maybe I'm being nitpicky here.

File: charts/prometheus-blackbox-exporter/templates/serviceaccount.yaml [hunk 5]
```diff
@@ -2,12 +2,11 @@
...
-    app.kubernetes.io/name: {{ include "prometheus-blackbox-exporter.name" . }}
-    app.kubernetes.io/instance: {{ .Release.Name }}
-    app.kubernetes.io/managed-by: {{ .Release.Service }}
-    helm.sh/chart: {{ include "prometheus-blackbox-exporter.chart" . }}
+    {{- include "prometheus-blackbox-exporter.labels" . | nindent 4 }}
+  {{- with .Values.serviceAccount.annotations }}
   annotations:
-{{ toYaml .Values.serviceAccount.annotations | indent 4 }}
-{{- end -}}
+    {{- toYaml . | nindent 4 }}
```
Comment (desaintmartin) [inline charts/prometheus-blackbox-exporter/templates/serviceaccount.yaml:10]: It depends what policy we want:
- all `{{ something }}` not indented
- follow logical indentation

If first:
```suggestion
{{ toYaml . | indent 4 }}
```

**Review by desaintmartin — COMMENTED** [review]
Nice re-write!

File: charts/prometheus-blackbox-exporter/templates/ingress.yaml (hunk 1 above)
Comment (monotek) [reply charts/prometheus-blackbox-exporter/templates/ingress.yaml:26]: Not sure if i get your point. 
ingressClassName is not used if kubeversion >=1.18-0 even if .Values.ingress.className is set.
So thats already what you want, or not?

File: charts/prometheus-blackbox-exporter/templates/ingress.yaml (hunk 2 above)
Comment (monotek) [reply charts/prometheus-blackbox-exporter/templates/ingress.yaml:29]: I've copied the default which is used when created with "helm create".
Do you see a benfit in using "with" as we would need to explicitly use the $ scope in the range using it?

File: charts/prometheus-blackbox-exporter/templates/service.yaml (hunk 3 above)
Comment (monotek) [reply charts/prometheus-blackbox-exporter/templates/service.yaml:6]: Woops... No...

File: charts/prometheus-blackbox-exporter/templates/serviceaccount.yaml (hunk 5 above)
Comment (monotek) [reply charts/prometheus-blackbox-exporter/templates/serviceaccount.yaml:10]: This comes also from "helm create" and imho it's nice using the same way as for the labels.
But i can live with the other version too :)

File: charts/prometheus-blackbox-exporter/templates/ingress.yaml (hunk 2 above)
Comment (desaintmartin) [reply charts/prometheus-blackbox-exporter/templates/ingress.yaml:29]: Not duplicating, but that is not really important.

File: charts/prometheus-blackbox-exporter/templates/ingress.yaml (hunk 1 above)
Comment (desaintmartin) [reply charts/prometheus-blackbox-exporter/templates/ingress.yaml:26]: My point is that if you set this value and deploy it in kube < 1.18, it will silently get ignored, it may be better to explicitely warn the user that it is incompatible.

File: charts/prometheus-blackbox-exporter/templates/serviceaccount.yaml (hunk 5 above)
Comment (desaintmartin) [reply charts/prometheus-blackbox-exporter/templates/serviceaccount.yaml:10]: let's stick with it then!

File: charts/prometheus-blackbox-exporter/templates/ingress.yaml (hunk 1 above)
Comment (monotek) [reply charts/prometheus-blackbox-exporter/templates/ingress.yaml:26]: I could add something like this to the NOTES.txt.

``` 
{{ if and .Values.ingress.className (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) }}
! WARNING !

You've set ".Values.ingressClassName" but it's not supported by your Kubernetes version!
Therefore option was not added in the ingress.yaml.
Please use 'kubernetes.io/ingress.class' annotation instead!
{{ end }}
```

Comment (monotek) [reply charts/prometheus-blackbox-exporter/templates/ingress.yaml:26]: @desaintmartin thoughts?

Comment (desaintmartin) [reply charts/prometheus-blackbox-exporter/templates/ingress.yaml:26]: Thanks! This way, it is explicit to the user and not silently swallowed.

**Review by desaintmartin — APPROVED** [review]

**Comment by desaintmartin** [comment]
lgtm!
Ready to merge? ;)

**Comment by onedr0p** [comment]
Anyone having issues installing this version of the chart?

Using just the default values it errors out:

```
Error: template: prometheus-blackbox-exporter/templates/servicemonitor.yaml:8:8: executing "prometheus-blackbox-exporter/templates/servicemonitor.yaml" at <include "prometheus-blackbox-exporter.labels" .>: error calling include: template: prometheus-blackbox-exporter/templates/_helpers.tpl:37:18: executing "prometheus-blackbox-exporter.labels" at <include "prometheus-blackbox-exporter.chart" .>: error calling include: template: prometheus-blackbox-exporter/templates/_helpers.tpl:30:25: executing "prometheus-blackbox-exporter.chart" at <.Chart.Name>: nil pointer evaluating interface {}.Name
```

Something with the labels got fudged up.

**Comment by onedr0p** [comment]
@monotek see above 👍🏼

**Comment by monotek** [comment]
Thanks for the info! 
Will be fixed with: https://github.com/prometheus-community/helm-charts/pull/1187
